File: Examples/ObjectiveCExampleApp/ObjectiveCExampleApp/Extensions/IASKAppSettingsViewController+Additions.m
#import "IASKAppSettingsViewController+Additions.h"
#import "Settings.h"

@implementation IASKAppSettingsViewController (Additions)

- (void)updateHiddenKeys {
    NSSet *hiddenKeys = [self computeHiddenKeysWithPriority:@[
        kIsPaymentSessionOnKey,
        kIsTokenAndSecretOnKey,
        kIsRecommendationOnKey,
        kIsAddressOnKey,
        kIsPrimaryAccountDetailsOnKey,
        kIsRecurringPaymentOnKey,
        kIsAutomaticReloadPaymentOnKey,
        kIsDeferredPaymentOnKey
    ]];
    [self setHiddenKeys:hiddenKeys];
}

- (void)didChangedSettingsWithKeys:(NSArray<NSString *> *)keys {

    NSSet *hiddenKeys;

    if ([keys containsObject:kIsPaymentSessionOnKey]) {
        hiddenKeys = [self computeHiddenKeysWithPriority:@[ kIsPaymentSessionOnKey, kIsAddressOnKey, kIsPrimaryAccountDetailsOnKey, kIsRecommendationOnKey, kIsRecurringPaymentOnKey ]];
    } else if ([keys containsObject:kIsTokenAndSecretOnKey]) {
        hiddenKeys = [self computeHiddenKeysWithPriority:@[ kIsTokenAndSecretOnKey, kIsAddressOnKey, kIsPrimaryAccountDetailsOnKey, kIsRecommendationOnKey, kIsRecurringPaymentOnKey ]];
    } else if ([keys containsObject:kIsAddressOnKey] || [keys containsObject:kIsPrimaryAccountDetailsOnKey] || [keys containsObject:kIsRecommendationOnKey] || [keys containsObject:kIsRecurringPaymentOnKey] || [keys containsObject:kIsAutomaticReloadPaymentOnKey] || [keys containsObject:kIsDeferredPaymentOnKey]) {
        hiddenKeys = [self computeHiddenKeysWithPriority:@[ kIsPaymentSessionOnKey, kIsTokenAndSecretOnKey, kIsRecommendationOnKey, kIsAddressOnKey, kIsPrimaryAccountDetailsOnKey, kIsRecurringPaymentOnKey, kIsAutomaticReloadPaymentOnKey, kIsDeferredPaymentOnKey ]];
    }

    if (hiddenKeys.count > 0) {
        [self setHiddenKeys:hiddenKeys animated:YES];
    }
}

- (NSSet *)computeHiddenKeysWithPriority:(NSArray *)keys {
    NSMutableArray *hiddenKeys = [NSMutableArray arrayWithArray:@[
        kTokenKey,
        kSecretKey,

        kSessionTokenKey,
        kPaymentSessionKey,
        kGeneratePaymentSessionKey,

        kRsaKey,
        kRecommendationUrlKey,
        kRecommendationTimeoutKey,
        kIsRecommendationHaltTransactionOnKey,

        kAddressLine1Key,
        kAddressLine2Key,
        kAddressLine3Key,
        kAddressTownKey,
        kAddressPostCodeKey,
        kAddressCountryCodeKey,
        kAddressAdministrativeDivisionKey,
        kAddressPhoneCountryCodeKey,
        kAddressMobileNumberKey,
        kAddressEmailAddressKey,

        kPrimaryAccountNameKey,
        kPrimaryAccountAccountNumberKey,
        kPrimaryAccountDateOfBirthKey,
        kPrimaryAccountPostCodeKey,

        kRecurringPaymentDescriptionKey,
        kRecurringPaymentManagementUrlKey,
        kRecurringPaymentBillingAgreementKey,
        kRecurringPaymentLabelKey,
        kRecurringPaymentAmountKey,
        kRecurringPaymentIntervalUnitKey,
        kRecurringPaymentIntervalCountKey,
        kRecurringPaymentStartDateKey,
        kRecurringPaymentEndDateKey,

        kAutomaticReloadPaymentDescriptionKey,
        kAutomaticReloadPaymentBillingAgreementKey,
        kAutomaticReloadPaymentManagementUrlKey,
        kAutomaticReloadPaymentLabelKey,
        kAutomaticReloadPaymentAmountKey,
        kAutomaticReloadPaymentThresholdAmountKey,

        kDeferredPaymentDescriptionKey,
        kDeferredPaymentBillingAgreementKey,
        kDeferredPaymentManagementUrlKey,
        kDeferredPaymentLabelKey,
        kDeferredPaymentAmountKey,
        kDeferredPaymentDeferredDateKey,
        kDeferredPaymentFreeCancellationDateKey
    ]];

    if ([keys containsObject:kIsPaymentSessionOnKey] && Settings.defaultSettings.isPaymentSessionAuthorizationOn) {
        [hiddenKeys removeObjectsInArray:@[ kSessionTokenKey, kPaymentSessionKey, kGeneratePaymentSessionKey]];
        [NSUserDefaults.standardUserDefaults setBool:NO forKey:kIsTokenAndSecretOnKey];
    }

    if ([keys containsObject:kIsTokenAndSecretOnKey] && Settings.defaultSettings.isTokenAndSecretAuthorizationOn) {
        [hiddenKeys removeObjectsInArray:@[ kTokenKey, kSecretKey ]];
        [NSUserDefaults.standardUserDefaults setBool:NO forKey:kIsPaymentSessionOnKey];
    }

    if ([keys containsObject:kIsRecommendationOnKey] && Settings.defaultSettings.isRecommendationFeatureOn) {
        [hiddenKeys removeObjectsInArray:@[
            kRsaKey,
            kRecommendationUrlKey,
            kRecommendationTimeoutKey,
            kIsRecommendationHaltTransactionOnKey,
        ]];
    }

    if ([keys containsObject:kIsAddressOnKey] && Settings.defaultSettings.isAddressOn) {
        [hiddenKeys removeObjectsInArray:@[
            kAddressLine1Key,
            kAddressLine2Key,
            kAddressLine3Key,
            kAddressTownKey,
            kAddressPostCodeKey,
            kAddressCountryCodeKey,
            kAddressAdministrativeDivisionKey,
            kAddressPhoneCountryCodeKey,
            kAddressMobileNumberKey,
            kAddressEmailAddressKey,
        ]];
    }

    if ([keys containsObject:kIsPrimaryAccountDetailsOnKey] && Settings.defaultSettings.isPrimaryAccountDetailsOn) {
        [hiddenKeys removeObjectsInArray:@[
            kPrimaryAccountNameKey,
            kPrimaryAccountAccountNumberKey,
            kPrimaryAccountDateOfBirthKey,
            kPrimaryAccountPostCodeKey
        ]];
    }

    if ([keys containsObject:kIsRecurringPaymentOnKey] && Settings.defaultSettings.isApplePayRecurringPaymentOn) {
        [hiddenKeys removeObjectsInArray:@[
            kRecurringPaymentDescriptionKey,
            kRecurringPaymentBillingAgreementKey,
            kRecurringPaymentManagementUrlKey,
            kRecurringPaymentLabelKey,
            kRecurringPaymentAmountKey,
            kRecurringPaymentIntervalUnitKey,
            kRecurringPaymentIntervalCountKey,
            kRecurringPaymentStartDateKey,
            kRecurringPaymentEndDateKey
        ]];
    }

    if ([keys containsObject:kIsAutomaticReloadPaymentOnKey] && Settings.defaultSettings.isApplePayAutomaticReloadPaymentOn) {
        [hiddenKeys removeObjectsInArray:@[
            kAutomaticReloadPaymentDescriptionKey,
            kAutomaticReloadPaymentBillingAgreementKey,
            kAutomaticReloadPaymentManagementUrlKey,
            kAutomaticReloadPaymentLabelKey,
            kAutomaticReloadPaymentAmountKey,
            kAutomaticReloadPaymentThresholdAmountKey
        ]];
    }

    if ([keys containsObject:kIsDeferredPaymentOnKey] && Settings.defaultSettings.isApplePayDeferredPaymentOn) {
        [hiddenKeys removeObjectsInArray:@[
            kDeferredPaymentDescriptionKey,
            kDeferredPaymentBillingAgreementKey,
            kDeferredPaymentManagementUrlKey,
            kDeferredPaymentLabelKey,
            kDeferredPaymentAmountKey,
            kDeferredPaymentDeferredDateKey,
            kDeferredPaymentFreeCancellationDateKey
        ]];
    }

    return [NSSet setWithArray:hiddenKeys];
}

@end

File: Examples/SwiftExampleApp/SwiftExampleApp/Extensions/IASKAppSettingsViewController+Additions.swift
import InAppSettingsKit

extension IASKAppSettingsViewController {
    func updateHiddenKeys() {
        let hiddenKeys = computeHiddenKeys(from: [kIsPaymentSessionOnKey,
                                                  kIsTokenAndSecretOnKey,
                                                  kIsAddressOnKey,
                                                  kIsRecommendationOnKey,
                                                  kIsPrimaryAccountDetailsOnKey,
                                                  kIsRecurringPaymentOnKey,
                                                  kIsDeferredPaymentOnKey])
        setHiddenKeys(hiddenKeys, animated: false)
    }

    func computeHiddenKeys(from keys: [String]) -> Set<String> {
        var hiddenKeys: Set<String> = [
            kTokenKey,
            kSecretKey,

            kSessionTokenKey,
            kPaymentSessionKey,

            kRsaKey,
            kRecommendationUrlKey,
            kRecommendationTimeoutKey,

            kAddressLine1Key,
            kAddressLine2Key,
            kAddressLine3Key,
            kAddressTownKey,
            kAddressPostCodeKey,
            kAddressCountryCodeKey,
            kAddressAdministrativeDivisionKey,
            kAddressPhoneCountryCodeKey,
            kAddressMobileNumberKey,
            kAddressEmailAddressKey,

            kPrimaryAccountNameKey,
            kPrimaryAccountAccountNumberKey,
            kPrimaryAccountDateOfBirthKey,
            kPrimaryAccountPostCodeKey,

            kRecurringPaymentDescriptionKey,
            kRecurringPaymentBillingAgreementKey,
            kRecurringPaymentManagementUrlKey,
            kRecurringPaymentLabelKey,
            kRecurringPaymentAmountKey,
            kRecurringPaymentIntervalUnitKey,
            kRecurringPaymentIntervalCountKey,
            kRecurringPaymentStartDateKey,
            kRecurringPaymentEndDateKey,

            kDeferredPaymentDescriptionKey,
            kDeferredPaymentBillingAgreementKey,
            kDeferredPaymentManagementUrlKey,
            kDeferredPaymentLabelKey,
            kDeferredPaymentAmountKey,
            kDeferredPaymentDeferredDateKey,
            kDeferredPaymentFreeCancellationDateKey
        ]

        if keys.contains(kIsPaymentSessionOnKey), Settings.standard.isSessionAuthorizationOn {
            hiddenKeys.remove(kSessionTokenKey)
            hiddenKeys.remove(kPaymentSessionKey)
            UserDefaults.standard.setValue(false, forKey: kIsTokenAndSecretOnKey)
        }

        if keys.contains(kIsTokenAndSecretOnKey), Settings.standard.isBasicAuthorizationOn {
            hiddenKeys.remove(kTokenKey)
            hiddenKeys.remove(kSecretKey)
            UserDefaults.standard.setValue(false, forKey: kIsPaymentSessionOnKey)
        }

        if keys.contains(kIsRecommendationOnKey), Settings.standard.isRecommendationOn {
            hiddenKeys.remove(kRsaKey)
            hiddenKeys.remove(kRecommendationUrlKey)
            hiddenKeys.remove(kRecommendationTimeoutKey)
        }

        if keys.contains(kIsAddressOnKey), Settings.standard.isAddressOn {
            hiddenKeys.remove(kAddressLine1Key)
            hiddenKeys.remove(kAddressLine2Key)
            hiddenKeys.remove(kAddressLine3Key)
            hiddenKeys.remove(kAddressTownKey)
            hiddenKeys.remove(kAddressPostCodeKey)
            hiddenKeys.remove(kAddressCountryCodeKey)
            hiddenKeys.remove(kAddressAdministrativeDivisionKey)
            hiddenKeys.remove(kAddressPhoneCountryCodeKey)
            hiddenKeys.remove(kAddressMobileNumberKey)
            hiddenKeys.remove(kAddressEmailAddressKey)
        }

        if keys.contains(kIsPrimaryAccountDetailsOnKey), Settings.standard.isPrimaryAccountDetailsOn {
            hiddenKeys.remove(kPrimaryAccountNameKey)
            hiddenKeys.remove(kPrimaryAccountAccountNumberKey)
            hiddenKeys.remove(kPrimaryAccountDateOfBirthKey)
            hiddenKeys.remove(kPrimaryAccountPostCodeKey)
        }

        if keys.contains(kIsRecurringPaymentOnKey), Settings.standard.isRecurringPaymentOn {
            hiddenKeys.remove(kRecurringPaymentDescriptionKey)
            hiddenKeys.remove(kRecurringPaymentBillingAgreementKey)
            hiddenKeys.remove(kRecurringPaymentManagementUrlKey)
            hiddenKeys.remove(kRecurringPaymentLabelKey)
            hiddenKeys.remove(kRecurringPaymentAmountKey)
            hiddenKeys.remove(kRecurringPaymentIntervalUnitKey)
            hiddenKeys.remove(kRecurringPaymentIntervalCountKey)
            hiddenKeys.remove(kRecurringPaymentStartDateKey)
            hiddenKeys.remove(kRecurringPaymentEndDateKey)
            UserDefaults.standard.setValue(false, forKey: kIsDeferredPaymentOnKey)
        }

        if keys.contains(kIsDeferredPaymentOnKey), Settings.standard.isDeferredPaymentOn {
            hiddenKeys.remove(kDeferredPaymentDescriptionKey)
            hiddenKeys.remove(kDeferredPaymentBillingAgreementKey)
            hiddenKeys.remove(kDeferredPaymentManagementUrlKey)
            hiddenKeys.remove(kDeferredPaymentLabelKey)
            hiddenKeys.remove(kDeferredPaymentAmountKey)
            hiddenKeys.remove(kDeferredPaymentDeferredDateKey)
            hiddenKeys.remove(kDeferredPaymentFreeCancellationDateKey)
            UserDefaults.standard.setValue(false, forKey: kIsRecurringPaymentOnKey)
        }

        return hiddenKeys
    }
}
